Street level photo from the Norwegian Public Road Administration

noenandre opened the pull request:
This PR tries to bring street level photos from the Norwegian Public Road Administration to iD.
The service is called "Vegbilder" which simply means road photos.

The Norwegian OSM community has obtained [permission ](https://forum.openstreetmap.org/viewtopic.php?pid=712312#p712312) to use all open services from the Norwegian road administration.

Every national and every county road is photographed bi-annual, and the service contains regular images and from the year 2021 also 360° image spheres, so this will be a great addition to iD for Norwegian users.

The PR still has some bugs and kinks to straiten out, and I do hope that iD delopers have some future capacity to eventually get this all over the finish line. 
One thing I don't want to tackle in this PR is the lack of a geo-filteing for street level photos overlays. Photos from the Norwegian Public Road Administration is of course only available in Norway, and id should only show "Vegbilder" as a provider when zoomed in there, similarly to aerial photos.

noenandre committed:
Initial work for new street level photo provider "Vegbilder" from the Norwegian Public Roads Administration.

noenandre committed:
Rename pannellum-streetside dist folder to pannellum, since Pannellum isn't used by just Bing Streetside anymore.
TODO: Pannellum is loaded twice.

noenandre committed:
Fetch available WFS image layers from OWS server. Use maps in cache, instead of objects.

noenandre committed:
Implemented image filtering.

noenandre committed:
Order sequences by time, primarily, and not linear reference.

noenandre committed:
Estimate image direction (azimuth) if missing.

noenandre committed:
Filter sequences too.

noenandre committed:
Split concern for panoramic and flat photos to own module.

noenandre committed:
Fix zoom for plane photos. TODO: set translateExtent correctly.

noenandre committed:
Removed a variable, renamed a property.

noenandre committed:
Test for init and reset methods on service.
Bugfix for reset.
Adjusted expectation of number of svg layers.

noenandre committed:
More info and localization in attribution bar.

noenandre committed:
Create sequences for each (wfs)layer instead of for all points at once. Rearrange cache.

noenandre commented:
Update:
A "Vegbilder" viewer need a viewer for both plane images and 360° images.
I have relied on the code for the Streetlevel viewer to make a viewer for 360° images, and Kartaview viewer to make my own viewer for plane photos. 
But the Kartaview viewer has some issues. It makes a new d3 zoom behaviour object on photoviewer resize but forgets to register/call it (#8997), and it uses the css property `object-fit: cover` which cut images that don't have the same ratio as the photo-viewer, when the desired behaviour is to hide the overflow, so the whole images is possible to pan over.
I have solved those issues, but also created a new one. 
Now the image won't zoom in anywhere near where the mousepointer is placed when using the scroll wheel, and it gets worse if you decrease the width of the photoviewer.

noenandre committed:
Attempt to fix plane photoviewer.

noenandre committed:
Create a new zoom behaviour when the image size potentially change.

noenandre committed:
Small bugfix.

noenandre committed:
Use last calculated missing heading for null vectors.

noenandre committed:
Tests for service/vegbilder is as comprehensive as similar providers.

noenandre committed:
Use nullish coalescing operator where it makes sense.

noenandre committed:
Split a promise into two.

noenandre committed:
Remove a global module variable "_availableLayers"; move information to cache.
Add new test "fetches available layers".

noenandre commented:
This PR is ready for review! I will be happy yo create a new PR if preferred.

tyrasd committed:
Merge branch 'develop' into vegbilder

tyrasd commented on modules/services/vegbilder.js:
```diff
@@ -0,0 +1,609 @@
+    }
+
+    this.event = utilRebind(this, dispatch, 'on');
+  },
```
I've noticed that this can cause an exception in [`renderer/photos.js`](https://github.com/openstreetmap/iD/blob/ddc9c6c67460e20a9531e6c759c61bebc5269b80/modules/renderer/photos.js#L209) (when opening a photo via URL parameter is requested).

I think we should not `await` the cache reset before binding the `on` method here:

 ```suggestion
  init: async function () {
    this.event = utilRebind(this, dispatch, 'on');
    if (!_vegbilderCache) {
      await this.reset();
    }
  },
```

But then `loadImages` needs to check if the cache is already populated:

```javascript
  loadImages: async function (context, margin) {
    margin ??= 1;
    if (!_vegbilderCache) {
      await this.reset();
    }
    const wfslayers = filterAvailableLayers(context.photos());
```

The same also in the `sequences` method plus some added async/await on top.

<hr>

Maybe there's a better option to fix this.

noenandre replied:
I have rearranged `init` as you suggested, and rearranged  `reset` to initialize a empty cache right away, before waiting for `fetchAvailableLayers`. Then it should not be necessary to check if the cache is there, in every function accessing the cache. Is that a better solution?

tyrasd reviewed (commented):
Hey. This is great!

Just a bit of first feedback from trying this out:

* Optimally, the layer should only show up and request/load layers if the user maps in the region this service covers. Similar to how the background layers and certain presets only show up if the user is mapping in the respective region.
* I've encountered a somewhat consistently thrown exception by `pannelum.js` (after being called by `_pannellumViewer.loadScene(key, pitch, yaw);` in `selectPhoto`)
  ````
  Uncaught TypeError: Cannot set properties of undefined (setting 'src')
    at FileReader.<anonymous> (pannellum.js:50:219)
  ````
  But only for the first 360°-pano to be displayed and it also does not seem to cause any noticeable issues. Did you encounter this as well?

noenandre committed:
Too set autoLoad to false when adding a scene,
seems too fix "Uncaught TypeError: Cannot set properties of undefined".

noenandre committed:
Reorder execution of init.
Set "on" binding before awaiting cache reset.
Set "wfslayers" to a empty map
immediately on reset.

noenandre commented:
> I've encountered a somewhat consistently thrown exception by pannelum.js (after being called by _pannellumViewer.loadScene(key, pitch, yaw); in selectPhoto)

Yes, I see that too, I don't know why I didn't catch that sooner. If I call `pannellum.addScene` without the option `autoLoad: true` the error seems to go away. `pannellum.loadScene` is called subsequently anyway.

> Optimally, the layer should only show up and request/load layers if the user maps in the region this service covers. Similar to how the background layers and certain presets only show up if the user is mapping in the respective region.

Yes, ideally it should. I am willing to work on such a feature, whenever this PR is accepted.

tyrasd committed:
Merge branch 'vegbilder' of github.com:noenandre/iD into vegbilder

tyrasd committed:
Merge branch 'develop' into vegbilder

tyrasd committed:
only show regional street-level photo providers in the available region

and disable selection checkbox if not zoomed in enough to load the photos

tyrasd committed:
only fetch vegbilder layers if they are needed

i.e. when enabling the respective street-side photo layer

tyrasd committed:
fix zoom/pan behavior on non-360° vegbilder photos

tyrasd committed:
show spinner while loading non-360° "vegbilder" photos

tyrasd committed:
reset yaw when switching between 360° and planar photo

(and also immediately when switching between different 360° photos)

tyrasd committed:
add to changelog

tyrasd reviewed (approved):
> the layer should only show up and request/load layers if the user maps in the region this service covers

d7902e138 and edb7c3b0f implement this.

While at it, I've also fixed a few issues I found (see additional commits in this branch), e.g. that zooming on non-360° photos was not preserving the center at the mouse location. Maybe you could double-check if I have not introduced any new bugs? :wink:

tyrasd committed:
fix test

noenandre commented:
The "plane" photo viewer works great now. Can't find any bugs, old or new.
